[git-webkit] Handle colliding branch names locally

JonWBedard opened the pull request:
#### b0f1b831ed3c5fa2efca2d65436dfb3e4f8605ac
<pre>
[git-webkit] Handle colliding branch names locally
<a href="https://bugs.webkit.org/show_bug.cgi?id=245444">https://bugs.webkit.org/show_bug.cgi?id=245444</a>
&lt;rdar://100184775&gt;

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.branches_for): Return an empty list for a non-existent remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.parser): Add --delete-existing flag.
(Branch.main): If we find an existing branch, rebase, delete or fail based on the --delete-existing flag.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pull_request_branch_point): Pass target_remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:

Canonical link: <a href="https://commits.webkit.org/254738@main">https://commits.webkit.org/254738@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ade5dde9c7484e72febd971acc0151d9a01af731

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34629 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/99406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/156910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33121 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/93701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95731 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/93663 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/89161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1419 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -169,10 +175,30 @@ def main(cls, args, repository, why=None, redact=False, **kwargs):
+                log.warning("Rebasing existing branch '{}' instead of creating a new one".format(args.issue))
+                if run([repository.executable(), 'rebase', 'HEAD', args.issue, '--autostash'], cwd=repository.root_path).returncode:
+                    sys.stderr.write("Failed to rebase '{}'\n".format(args.issue))
+                    return 1
```
If the rebase introduces conflicts, the rebase hasn't _failed_, it just needs manual fixups and a call to `git rebase --continue`. Can we avoid writing our own error message and count on git's output to be informative enough?
```suggestion
                if run([repository.executable(), 'rebase', 'HEAD', args.issue, '--autostash'], cwd=repository.root_path).returncode:
                    return 1
```

Saying "Failed to rebase" when there are conflicts and the original commits haven't been applied yet may make it seem like we have caused data loss, when in reality the local changes will come back after running `git rebase --continue`, and the original branch can be restored with `git rebase --abort`.

JonWBedard replied:
Will remove this error message, was trying to be helpful, but yeah, I can see how our editorializing here is not useful.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -169,10 +175,30 @@ def main(cls, args, repository, why=None, redact=False, **kwargs):
-            if branch == args.issue or remote_re.match(branch):
-                sys.stderr.write("'{}' already exists\n".format(args.issue))
+        if args.issue in repository.branches_for(remote=target_remote):
+            if not args.delete_existing:
+                sys.stderr.write("'{}' exists on the remote '{}' and will be overwritten by a push\n".format(args.issue, target_remote))
+                if args.delete_existing is False:
+                    return 1
```
--delete-existing's argument string implies that it only affects local branches. Can you change it to clarify what it does to remote branches too?

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -169,10 +175,30 @@ def main(cls, args, repository, why=None, redact=False, **kwargs):
+        for remote in repository.source_remotes():
+            if args.issue in repository.branches_for(remote=remote):
+                sys.stderr.write("'{}' cannot be used as a PR branch, it exists on remote '{}'\n".format(args.issue, remote))
+                return 1
```
What is this guarding against? I can't think of a reason why we'd compute a branch name that happens to match a branch name on origin, but that shouldn't _technically_ be unusable as a PR branch, right?

Even `main` is a valid PR branch!

emw-apple replied:
(Well, maybe a PR from `fork/main` would break EWS, but at least it's valid for Git and GitHub 😛)

JonWBedard replied:
I was guarding against things like `fork/main` and `fork/safari-613-branch`, but this is probably more trouble than it's worth the more I think about it, will remove this code.

JonWBedard committed:
[git-webkit] Handle colliding branch names locally
https://bugs.webkit.org/show_bug.cgi?id=245444
<rdar://100184775>

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.branches_for): Return an empty list for a non-existent remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.parser): Add --delete-existing flag.
(Branch.main): If we find an existing branch, rebase, delete or fail based on the --delete-existing flag.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pull_request_branch_point): Pass target_remote.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:

Canonical link: https://commits.webkit.org/254738@main
